### PR TITLE
IR-116: Return sprint summary in sprint Get API

### PR DESCRIPTION
**Pull request by ishugupta92**

### CODE REVIEW

File: apps/retrospective/serializers/validators/retrospective.go [hunk 1]
```diff
@@ -35,9 +35,25 @@ func IsValidTaskProviderConfigList(
 		if providerType, isValid = taskProviderConfig["type"].(string); !isValid {
 			return false
 		}
+
 		if providerData, isValid = taskProviderConfig["data"].(map[string]interface{}); !isValid {
 			return false
 		}
+
+		for _, taskType := range []string{"FeatureTypes", "TaskTypes", "BugTypes"} {
```
Comment (anmolghosh) [inline apps/retrospective/serializers/validators/retrospective.go]: make this array a contstant

File: apps/tasktracker/providers.go [hunk 2]
```diff
@@ -183,6 +184,51 @@ func GetConnections(config []byte) (connections []Connection, err error) {
 	return connections, nil
 }
 
+// GetTaskTypeMappings ...
+func GetTaskTypeMappings(config []byte) (featureTypes []string, taskTypes []string, bugTypes []string, err error) {
```
Comment (anmolghosh) [inline apps/tasktracker/providers.go]: use single struct to return these

File: apps/retrospective/serializers/sprint.go [hunk 3]
```diff
@@ -8,16 +8,41 @@ import (
...
+	MemberCount      int
+	TotalAllocation  float64
+	TotalExpectation float64
+	Holidays         float64
+	TotalVacations   float64
+	TargetSP         float64
+	TaskTypes        SprintTaskTypesSummary
+}
+
+type SprintTaskTypesSummary struct {
```
Comment (anmolghosh) [inline apps/retrospective/serializers/sprint.go]: As discussed, this could be dynamic

**Review by anmolghosh — CHANGES_REQUESTED** [review]

File: apps/retrospective/serializers/validators/retrospective.go (hunk 1 above)
Comment (ishugupta92) [reply apps/retrospective/serializers/validators/retrospective.go]: Done

File: apps/tasktracker/providers.go (hunk 2 above)
Comment (ishugupta92) [reply apps/tasktracker/providers.go]: Done

File: apps/retrospective/serializers/sprint.go (hunk 3 above)
Comment (ishugupta92) [reply apps/retrospective/serializers/sprint.go]: Done

File: apps/retrospective/services/sprint.go [hunk 4]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+		summary[taskType], status, err = service.getSprintTaskTypeQuery(sprintID, taskTypes[taskType])
+		if err != nil {
+			return summary, status, err
+		}
+	}
+
+	return summary, http.StatusOK, nil
+}
+
+func (service SprintService) getSprintTaskTypeQuery(sprintID string, taskTypes string) (summary retroSerializers.SprintTaskTypeSummary, status int, err error) {
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: Is the method returning data or query?

File: apps/retrospective/services/sprint.go [hunk 5]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+// GetSprintSummary ...
+func (service SprintService) GetSprintSummary(sprintID string, retroID uint) (summary retroSerializers.SprintSummary, status int, err error) {
+	db := service.DB
+
+	var sprint retroModels.Sprint
+
+	err = db.Model(&retroModels.Sprint{}).Where("id = ?", sprintID).Preload("Retrospective").First(&sprint).Error
+	if err != nil {
+		if err == gorm.ErrRecordNotFound {
+			return summary, http.StatusNotFound, errors.New("sprint not found")
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/services/sprint.go [hunk 6]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+		}
+		utils.LogToSentry(err)
+		return summary, http.StatusInternalServerError, errors.New("failed to get sprint summary")
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/services/sprint.go [hunk 7]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+		Select("COUNT(*) as member_count, " +
+			"SUM(allocation_percent) as total_allocation, " +
+			"SUM(expectation_percent) as total_expectation, " +
+			"sum(vacations) as total_vacations," +
+			"0 as holidays").
+		Scan(&summary).Error
+
+	if err != nil {
+		utils.LogToSentry(err)
+		return summary, http.StatusInternalServerError, errors.New("failed to get sprint summary")
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/services/sprint.go [hunk 8]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+	}
+
+	summary.TargetSP = utils.CalculateExpectedSP(*sprint.StartDate, *sprint.EndDate, summary.TotalVacations / float64(summary.MemberCount),
+		summary.TotalAllocation, summary.TotalExpectation, sprint.Retrospective.StoryPointPerWeek)
+
+	taskTypesSummary, status, err := service.GetSprintTaskTypesSummary(sprintID, retroID)
+
+	summary.TaskTypes = taskTypesSummary
+	if err != nil {
+		return summary, status, errors.New("failed to get sprint")
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/services/sprint.go [hunk 9]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+	if err != nil {
+		utils.LogToSentry(err)
+		return summary, http.StatusInternalServerError, errors.New("failed to get sprint summary")
+	}
+	summary = make(map[string]retroSerializers.SprintTaskTypeSummary)
+
+	for _, taskType := range tasktracker.TaskTypes {
+		summary[taskType], status, err = service.getSprintTaskTypeQuery(sprintID, taskTypes[taskType])
+		if err != nil {
+			return summary, status, err
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/serializers/sprint.go [hunk 10]
```diff
@@ -8,16 +8,35 @@ import (
...
+}
+
+type SprintSummary struct {
+	MemberCount      int
+	TotalAllocation  float64
+	TotalExpectation float64
+	Holidays         float64
+	TotalVacations   float64
+	TargetSP         float64
+	TaskTypes        map[string]SprintTaskTypeSummary
```
Comment (anmolghosh) [inline apps/retrospective/serializers/sprint.go]: `TaskTypesSummary` or `TaskSummary` is more apt.
My pred is `TaskSummary`. Please take call

File: apps/retrospective/serializers/sprint.go [hunk 11]
```diff
@@ -8,16 +8,35 @@ import (
...
+}
+
+type SprintTaskTypeSummary struct {
```
Comment (anmolghosh) [inline apps/retrospective/serializers/sprint.go]: do we need to add type here in the struct name? let's discuss!

File: apps/retrospective/services/sprint.go [hunk 12]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+	summary.TaskTypes = taskTypesSummary
+	if err != nil {
+		return summary, status, errors.New("failed to get sprint")
+	}
+
+	return summary, http.StatusOK, nil
+}
+
+// GetSprintTaskTypesSummary ...
+func (service SprintService) GetSprintTaskTypesSummary(sprintID string, retroID uint) (summary map[string]retroSerializers.SprintTaskTypeSummary, status int, err error) {
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: let's discuss the naming

File: apps/retrospective/services/sprint.go [hunk 13]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+	db := service.DB
+	var retro retroModels.Retrospective
+
+	err = db.Model(&retroModels.Retrospective{}).Where("id = ?", retroID).First(&retro).Error
+	if err != nil {
+		if err == gorm.ErrRecordNotFound {
+			return summary, http.StatusNotFound, errors.New("retrospective not found")
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/services/sprint.go [hunk 14]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+		}
+		utils.LogToSentry(err)
+		return summary, http.StatusInternalServerError, errors.New("failed to get sprint summary")
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/services/sprint.go [hunk 15]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+	}
+
+	taskTypes, err := tasktracker.GetTaskTypeMappings(retro.TaskProviderConfig)
+
+	if err != nil {
+		utils.LogToSentry(err)
+		return summary, http.StatusInternalServerError, errors.New("failed to get sprint summary")
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: shouldn't we return nil instead of summary

File: apps/retrospective/services/sprint.go [hunk 16]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+		Where("sprints.end_date >= tasks.done_at").
+		Select("COUNT(*) as count, COALESCE(SUM(sprint_member_tasks.points_earned),0) as points_earned").
+		QueryExpr()
+
+	taskQuery := taskList.Select("COUNT(*) as total_count, COALESCE(SUM(sprint_member_tasks.points_earned),0) as total_points_earned").QueryExpr()
+	err = db.Raw("SELECT * FROM (?) as total CROSS JOIN (?) as done", taskQuery, doneTaskQuery).Scan(&summary).Error
+
+	if err != nil {
+		utils.LogToSentry(err)
+		return summary, status, err
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: should append task type in error message for more clarity

Comment (anmolghosh) [reply apps/retrospective/services/sprint.go]: what is value of status here ?

File: apps/retrospective/services/sprint.go [hunk 17]
```diff
@@ -127,19 +127,129 @@ func (service SprintService) Get(sprintID string) (*retroSerializers.Sprint, int
...
+	db := service.DB
+
+	taskList := db.Model(&retroModels.Sprint{}).
+		Scopes(retroModels.SprintJoinSM, retroModels.SMJoinSMT, retroModels.SMTJoinTask).
+		Where("sprints.id = ?", sprintID).
+		Where("LOWER(tasks.type) in (?)", taskTypes)
+
+	doneTaskQuery := taskList.Where("sprints.start_date <= tasks.done_at").
+		Where("sprints.end_date >= tasks.done_at").
+		Select("COUNT(*) as count, COALESCE(SUM(sprint_member_tasks.points_earned),0) as points_earned").
```
Comment (anmolghosh) [inline apps/retrospective/services/sprint.go]: use capital letter for query keywords such as `AS`, `IN`, etc

**Review by anmolghosh — CHANGES_REQUESTED** [review]
Please review the comments and Fix codecacy

File: apps/retrospective/services/sprint.go (hunk 4 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: It is returning data. I changed the name.

File: apps/retrospective/services/sprint.go (hunk 5 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/services/sprint.go (hunk 6 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/services/sprint.go (hunk 7 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/services/sprint.go (hunk 8 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/services/sprint.go (hunk 9 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/serializers/sprint.go (hunk 10 above)
Comment (ishugupta92) [reply apps/retrospective/serializers/sprint.go]: Changed to `TaskSummary`

File: apps/retrospective/serializers/sprint.go (hunk 11 above)
Comment (ishugupta92) [reply apps/retrospective/serializers/sprint.go]: Removed

File: apps/retrospective/services/sprint.go (hunk 12 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Changed as discussed

File: apps/retrospective/services/sprint.go (hunk 13 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/services/sprint.go (hunk 14 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/services/sprint.go (hunk 15 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

File: apps/retrospective/services/sprint.go (hunk 16 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done
Changed status to `500`

File: apps/retrospective/services/sprint.go (hunk 17 above)
Comment (ishugupta92) [reply apps/retrospective/services/sprint.go]: Done

**Review by anmolghosh — APPROVED** [review]

**Pull request merged**